packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,239 @@
+                )
+
+                if (entryMod.resource) {
+                  entryNameMap.set(entryMod.resource, name)
+                  entryModMap.set(entryMod.resource, entryMod)
+                }
+              }
+            })
+
+            // TODO: investigate allowing non-sync fs calls in node-file-trace
+            // for better performance
+            const readFile = (path: string) => {
+              const mod = depModMap.get(path) || entryModMap.get(path)
+
+              // map the transpiled source when available to avoid
+              // parse errors in node-file-trace
+              const source = mod?.originalSource()
+
+              if (source?._valueAsBuffer) {
+                return source._valueAsBuffer
+              }
+
+              try {
+                return compilation.inputFileSystem.readFileSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT' || e.code === 'EISDIR') {
+                  return null
+                }
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
+              } catch (e) {
+                if (
+                  e.code !== 'EINVAL' &&
+                  e.code !== 'ENOENT' &&
+                  e.code !== 'UNKNOWN'
+                ) {
+                  throw e
+                }
+                return null
+              }
+            }
+            const stat = (path: string) => {
+              try {
+                return compilation.inputFileSystem.statSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT') {
+                  return null
+                }
+                throw e
+              }
+            }
+
+            const nftCache = {}
+            const entryPaths = Array.from(entryModMap.keys())
+
+            for (const entry of entryPaths) {
+              depModMap.clear()
+              const entryMod = entryModMap.get(entry)
+              const cachedTraces = entryMod.buildInfo?.cachedNextEntryTrace
+
+              // Use cached trace if available and trace version matches
+              if (
+                isWebpack5 &&
+                cachedTraces &&
+                cachedTraces.version === TRACE_OUTPUT_VERSION
+              ) {
+                this.entryTraces.set(
+                  entryNameMap.get(entry)!,
+                  cachedTraces.tracedDeps
+                )
+                continue
+              }
+
+              for (const dep of entryMod.dependencies) {
+                const depMod = getModuleFromDependency(compilation, dep)
+
+                if (depMod?.resource) {
+                  depModMap.set(depMod.resource, depMod)
+                }
+              }
+
+              const toTrace: string[] = [entry, ...depModMap.keys()]
+
+              const root = nodePath.parse(process.cwd()).root
```

Comment:
```suggestion
              const root = compiler.options.context
```

Reply:
Why is the root different than `this.appDir`. Is this because we want to trace some files above the app dir like a monorepo?

Reply:
Yeah exactly